### PR TITLE
perf(es/lints): Avoid needless allocations in `no-dupe-args`

### PR DESCRIPTION
**Pull request by kdy1**
**Description:**

I introduced a zero-allocation variant for `swc_ecma_utils::DestructuringFinder` that does not allocate anything.

**Related issue (if exists):**

### CODE REVIEW

**Review by kdy1 — COMMENTED** [review]
swc-bump:
 - swc_ecma_lints
 - swc_ecma_utils

File: crates/swc_ecma_lints/src/rules/no_dupe_args.rs [hunk 1]
```diff
@@ -13,54 +12,65 @@ pub fn no_dupe_args() -> Box<dyn Rule> {
...
-                        .emit();
-                });
-            }
-        });
-    }
+/// This has time complexity of O(n^2), but it's fine as the number of paramters
+/// is usually small.
+macro_rules! check {
+    ($node:expr) => {{
+        // This allocates only if there are duplicate parameters.
```
Comment (sokra) [inline crates/swc_ecma_lints/src/rules/no_dupe_args.rs:19]: I don't think that's true...

File: crates/swc_ecma_lints/src/rules/no_dupe_args.rs [hunk 2]
```diff
@@ -13,54 +12,65 @@ pub fn no_dupe_args() -> Box<dyn Rule> {
...
+
+        let mut i1 = 0;
+        for_each_binding_ident($node, |id1| {
+            i1 += 1;
+
+            let mut i2 = 0;
+            for_each_binding_ident($node, |id2| {
+                i2 += 1;
+
+                if i1 >= i2 || done.contains(&i1) {
```
Comment (sokra) [inline crates/swc_ecma_lints/src/rules/no_dupe_args.rs:30]: The `Vec::contains` makes the whole algorithm `O(n^3)`

File: crates/swc_ecma_lints/src/rules/no_dupe_args.rs [hunk 3]
```diff
@@ -13,54 +12,65 @@ pub fn no_dupe_args() -> Box<dyn Rule> {
...
-                        )
-                        .span_label(old_span, "previous definition here".to_string())
-                        .span_label(span, &"used as parameter more than once".to_string())
-                        .emit();
-                });
-            }
-        });
-    }
+/// This has time complexity of O(n^2), but it's fine as the number of paramters
+/// is usually small.
```
Comment (sokra) [inline crates/swc_ecma_lints/src/rules/no_dupe_args.rs:16]: This probably holds true for 99.9%, but maybe we can add another implementation that is used for n > 100 based on a hashmap.

File: crates/swc_ecma_lints/src/rules/no_dupe_args.rs [hunk 4]
```diff
@@ -13,54 +12,65 @@ pub fn no_dupe_args() -> Box<dyn Rule> {
...
+
+            let mut i2 = 0;
+            for_each_binding_ident($node, |id2| {
+                i2 += 1;
+
+                if i1 >= i2 || done.contains(&i1) {
+                    return;
+                }
+
+                done.push(i1);
```
Comment (sokra) [inline crates/swc_ecma_lints/src/rules/no_dupe_args.rs:34]: I don't really get what this should do...

File: crates/swc_ecma_lints/src/rules/no_dupe_args.rs (hunk 1 above)
Comment (kdy1) [reply crates/swc_ecma_lints/src/rules/no_dupe_args.rs:19]: This comment was about `done,` but well I made a mistake.

File: crates/swc_ecma_lints/src/rules/no_dupe_args.rs [hunk 5]
```diff
@@ -13,54 +12,65 @@ pub fn no_dupe_args() -> Box<dyn Rule> {
...
+
+                if id1.span.ctxt == id2.span.ctxt && id1.sym == id2.sym {
+                    HANDLER.with(|handler| {
+                        handler
+                            .struct_span_err(
+                                id2.span,
+                                &format!(
+                                    "the name `{}` is bound more than once in this parameter list",
```
Comment (sokra) [inline crates/swc_ecma_lints/src/rules/no_dupe_args.rs:42]: If a binding is defined 3 times this error will be reported multiple times. That was already the case with the old implementation.

File: crates/swc_ecma_lints/src/rules/no_dupe_args.rs (hunk 4 above)
Comment (kdy1) [reply crates/swc_ecma_lints/src/rules/no_dupe_args.rs]: This was a mistake. It's now fixed.

File: crates/swc_ecma_lints/src/rules/no_dupe_args.rs (hunk 5 above)
Comment (kdy1) [reply crates/swc_ecma_lints/src/rules/no_dupe_args.rs:42]: That's handled by `done`. Actually I added `done` just to match the original behavior.

File: crates/swc_ecma_lints/src/rules/no_dupe_args.rs (hunk 2 above)
Comment (kdy1) [reply crates/swc_ecma_lints/src/rules/no_dupe_args.rs:30]: Now `done` is empty for almost all cases.

**Pull request merged**